pkg/resource/instance_profile/hooks.go

```diff
@@ -0,0 +1,239 @@
+	toDelete := []*svcapitypes.Tag{}
+	existingTags := latest.ko.Spec.Tags
+
+	for _, t := range existingTags {
+		if !inTags(*t.Key, *t.Value, desired.ko.Spec.Tags) {
+			toDelete = append(toDelete, t)
+		}
+	}
+	for _, t := range desired.ko.Spec.Tags {
+		if !inTags(*t.Key, *t.Value, existingTags) {
+			toAdd = append(toAdd, t)
+		}
+	}
+
+	if len(toDelete) > 0 {
+		for _, t := range toDelete {
+			rlog.Debug("removing tag from instance profile", "key", *t.Key, "value", *t.Value)
+		}
+		if err = rm.removeTags(ctx, desired, toDelete); err != nil {
+			return err
+		}
+	}
+	if len(toAdd) > 0 {
+		for _, t := range toAdd {
+			rlog.Debug("adding tag to instance profile", "key", *t.Key, "value", *t.Value)
+		}
+		if err = rm.addTags(ctx, desired, toAdd); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// inTags returns true if the supplied key and value can be found in the
+// supplied list of Tag structs.
+func inTags(
+	key string,
+	value string,
+	tags []*svcapitypes.Tag,
+) bool {
+	for _, t := range tags {
+		if *t.Key == key && *t.Value == value {
+			return true
+		}
+	}
+	return false
+}
+
+// addTags adds the supplied Tags to the supplied InstanceProfile resource
+func (rm *resourceManager) addTags(
+	ctx context.Context,
+	r *resource,
+	tags []*svcapitypes.Tag,
+) (err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.addTag")
+	defer func() { exit(err) }()
+
+	input := &svcsdk.TagInstanceProfileInput{}
+	input.InstanceProfileName = r.ko.Spec.Name
+	inTags := []*svcsdk.Tag{}
+
+	for _, t := range tags {
+		inTags = append(inTags, &svcsdk.Tag{Key: t.Key, Value: t.Value})
+	}
+	input.Tags = inTags
+
+	_, err = rm.sdkapi.TagInstanceProfileWithContext(ctx, input)
+	rm.metrics.RecordAPICall("UPDATE", "TagInstanceProfile", err)
+	return err
+}
+
+// removeTags removes the supplied Tags from the supplied InstanceProfile resource
+func (rm *resourceManager) removeTags(
+	ctx context.Context,
+	r *resource,
+	tags []*svcapitypes.Tag,
+) (err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.removeTag")
+	defer func() { exit(err) }()
+
+	input := &svcsdk.UntagInstanceProfileInput{}
+	input.InstanceProfileName = r.ko.Spec.Name
+	inTagKeys := []*string{}
+
+	for _, t := range tags {
+		inTagKeys = append(inTagKeys, t.Key)
+	}
+	input.TagKeys = inTagKeys
+
+	_, err = rm.sdkapi.UntagInstanceProfileWithContext(ctx, input)
+	rm.metrics.RecordAPICall("UPDATE", "UntagInstanceProfile", err)
+	return err
+}
+
+// compareTags is a custom comparison function for comparing lists of Tag
+// structs where the order of the structs in the list is not important.
+func compareTags(
+	delta *ackcompare.Delta,
+	a *resource,
+	b *resource,
+) {
+	if len(a.ko.Spec.Tags) != len(b.ko.Spec.Tags) {
+		delta.Add("Spec.Tags", a.ko.Spec.Tags, b.ko.Spec.Tags)
+	} else if len(a.ko.Spec.Tags) > 0 {
+		if !commonutil.EqualTags(a.ko.Spec.Tags, b.ko.Spec.Tags) {
+			delta.Add("Spec.Tags", a.ko.Spec.Tags, b.ko.Spec.Tags)
+		}
+	}
+}
```

Comment:
Most of the tag code was honestly copied over...

Reply:
We do that all the time here :)